nav2_costmap_2d/src/costmap_2d_ros.cpp

```diff
@@ -176,6 +165,20 @@ Costmap2DROS::Costmap2DROS(const std::string & name, tf2_ros::Buffer & tf)
     {"transform_tolerance", "update_frequency", "publish_frequency", "width", "height",
     "resolution", "origin_x", "origin_y", "footprint_padding", "robot_radius", "footprint"});
   dynamic_param_client_->set_callback(std::bind(&Costmap2DROS::reconfigureCB, this));
+
+  // Set these parameters *after* the dynamic params stuff is set up
+  // so that the reconfigure callback is called
```

Comment:
There is already an interface to have the reconfigure callback called: `set_callback(callback, bool init_callback)` which force triggers a callback and is by default true. Thus, you can add parameters anytime to fill the internal parameter map before you decide to set the user callback on the `DynamicParamsClient` object. I need to update the README to reflect the newer interface. I'd suggest you'd want to set the parameters before so that they can be looked up on the node when they are being added via `add_parameter(...)`, otherwise each new `set_parameter_if_not_set` will trigger a new callback where the other parameter values won't have been set untill the last one.  If there is a desire to not set a parameter before, I'd recommend inside the callback to use the `get_event_param_or(...)` call with sensible values.

Reply:
@bpwilcox - thanks for the feedback. We could change this further if there's a better way to do it, but I'd request to do that in a follow on PR. This seems to work and gets us unblocked right now. I can file an issue as an enhancement to this.

Reply:
I suppose my point was that the callback should already be triggered as is in the original code before this PR (unless you discovered specific behavior that indicates otherwise, in which case go ahead), which seems to be the reason for the proposed change in this section. Consequently, this now adds redundant calls to the callback (which will also call `updateMap` for every callback, so a bit inefficient to do in the constructor). We could do a PR to revert later, but it seems simple enough to not fix if it isn't broken :)

Reply:
@bpwilcox - I believe it is broken. Before this change the node seems to get into an infinite callback loop. This was a fix @mjeronimo made and it fixed that problem. I'll open an issue for you to look at it when you get back, and if possible revert it. I'm open to a better solution.

Reply:
I added #352 for this issue.

Reply:
It doesn't make sense to me that this fixes the problem. These parameters should have their defaults set before the callback is called, otherwise there won't be a default value.